File: platform/workers/quant-ops-mcp/src/projection_content.js
/** Content verification for immutable Ops Projection payload tables. */

import {
  canonicalProjectionJson,
  projectionSha256,
} from "./projection_signature.js";

export const PROJECTED_CONTENT_TABLES = Object.freeze([
  "collection_sla_status",
  "coverage_segments",
  "dataset_coverage",
  "endpoint_inventory",
  "ingestion_run_log",
  "ingestion_validation",
  "ingestion_watermarks",
  "ops_alerts",
  "ops_b0_status",
  "ops_projection_metadata",
  "ops_ready_snapshots",
  "ops_ready_state",
  "ops_snapshot_quality",
  "ops_storage_plane_status",
  "ops_sync_feed",
  "raw_retention_manifests",
]);

const PROJECTED_CONTENT_TABLE_SET = new Set(PROJECTED_CONTENT_TABLES);

/** @param {Uint8Array} left @param {Uint8Array} right */
function compareBytes(left, right) {
  const length = Math.min(left.length, right.length);
  for (let index = 0; index < length; index += 1) {
    if (left[index] !== right[index]) return left[index] - right[index];
  }
  return left.length - right.length;
}

/** @param {unknown[]} rows */
function canonicalRows(rows) {
  const encoder = new TextEncoder();
  return rows
    .map((row) => ({ bytes: encoder.encode(canonicalProjectionJson(row)), row }))
    .sort((left, right) => compareBytes(left.bytes, right.bytes))
    .map(({ row }) => row);
}

/** @param {unknown} value */
function isDigest(value) {
  return typeof value === "string" && /^sha256:[0-9a-f]{64}$/.test(value);
}

/**
 * @param {D1Database} db
 * @param {string} generation
 * @param {string} table
 */
export async function projectedTableContent(db, generation, table) {
  if (!PROJECTED_CONTENT_TABLE_SET.has(table)) {
    throw new RangeError(`unknown Ops Projection content table: ${table}`);
  }
  const result = await db.prepare(
    `SELECT * FROM ${table} WHERE projection_generation_id=?`,
  ).bind(generation).all();
  const rows = canonicalRows(result.results || []);
  return {
    row_count: rows.length,
    content_digest: await projectionSha256({ rows }),
  };
}

/** @param {Record<string, {row_count:number,content_digest:string}>} manifest */
export async function projectedManifestDigest(manifest) {
  return projectionSha256({ tables: manifest });
}

/**
 * Verify that the signed manifest is complete and that every table used by a
 * tool still has exactly the rows the publisher sealed. The overall digest
 * binds all table digests even when one tool needs only a bounded subset.
 *
 * @param {D1Database} db
 * @param {Record<string, unknown>} envelope
 * @param {readonly string[]} requiredTables
 * @returns {Promise<{ok:boolean,reason:string|null}>}
 */
export async function verifyProjectedContent(db, envelope, requiredTables) {
  const manifestValue = envelope.content_manifest;
  const rowCountsValue = envelope.row_counts;
  if (!manifestValue || typeof manifestValue !== "object" || Array.isArray(manifestValue) ||
      !rowCountsValue || typeof rowCountsValue !== "object" || Array.isArray(rowCountsValue)) {
    return { ok: false, reason: "signed Ops Projection content manifest is missing" };
  }
  const manifest = /** @type {Record<string, unknown>} */ (manifestValue);
  const rowCounts = /** @type {Record<string, unknown>} */ (rowCountsValue);
  if (Object.keys(manifest).sort().join("\0") !== PROJECTED_CONTENT_TABLES.join("\0") ||
      Object.keys(rowCounts).sort().join("\0") !== PROJECTED_CONTENT_TABLES.join("\0")) {
    return { ok: false, reason: "signed Ops Projection content manifest membership drift" };
  }
  /** @type {Record<string, {row_count:number,content_digest:string}>} */
  const normalized = {};
  for (const table of PROJECTED_CONTENT_TABLES) {
    const raw = manifest[table];
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
      return { ok: false, reason: `signed Ops Projection manifest is malformed for ${table}` };
    }
    const row = /** @type {Record<string, unknown>} */ (raw);
    if (Object.keys(row).sort().join("\0") !== "content_digest\0row_count" ||
        !Number.isInteger(row.row_count) || Number(row.row_count) < 0 ||
        !isDigest(row.content_digest) || rowCounts[table] !== row.row_count) {
      return { ok: false, reason: `signed Ops Projection manifest is malformed for ${table}` };
    }
    normalized[table] = {
      row_count: Number(row.row_count),
      content_digest: String(row.content_digest),
    };
  }
  if (await projectedManifestDigest(normalized) !== envelope.content_digest) {
    return { ok: false, reason: "signed Ops Projection content digest does not bind its manifest" };
  }
  const uniqueTables = [...new Set(requiredTables)].sort();
  if (uniqueTables.some((table) => !PROJECTED_CONTENT_TABLE_SET.has(table))) {
    return { ok: false, reason: "tool requested an unknown Ops Projection content table" };
  }
  try {
    for (const table of uniqueTables) {
      const observed = await projectedTableContent(db, String(envelope.generation_id), table);
      const expected = normalized[table];
      if (observed.row_count !== expected.row_count ||
          observed.content_digest !== expected.content_digest) {
        return { ok: false, reason: `active Ops Projection content mismatch for ${table}` };
      }
    }
  } catch {
    return { ok: false, reason: "active Ops Projection content verification failed" };
  }
  return { ok: true, reason: null };
}

File: packages/data_plane/ops/projection_content.py
"""Canonical content addressing for immutable Ops Projection payload rows."""

from __future__ import annotations

import hashlib
import json
import math
from typing import Any, Iterable, Mapping


PROJECTED_CONTENT_TABLES = (
    "collection_sla_status",
    "coverage_segments",
    "dataset_coverage",
    "endpoint_inventory",
    "ingestion_run_log",
    "ingestion_validation",
    "ingestion_watermarks",
    "ops_alerts",
    "ops_b0_status",
    "ops_projection_metadata",
    "ops_ready_snapshots",
    "ops_ready_state",
    "ops_snapshot_quality",
    "ops_storage_plane_status",
    "ops_sync_feed",
    "raw_retention_manifests",
)


def canonical_content_bytes(value: Any) -> bytes:
    """Match the Worker canonicalizer: sorted keys, UTF-8, no whitespace."""

    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def projection_content_digest(value: Any) -> str:
    return "sha256:" + hashlib.sha256(canonical_content_bytes(value)).hexdigest()


def _storage_value(value: Any) -> Any:
    """Normalize values to SQLite/D1's JSON-visible storage representation."""

    if isinstance(value, bool):
        return int(value)
    if isinstance(value, float):
        if (
            not math.isfinite(value)
            or not value.is_integer()
            or not (-(2**63) <= value < 2**63)
        ):
            raise ValueError(
                "Ops Projection payloads do not permit non-integral REAL values"
            )
        return int(value)
    if isinstance(value, Mapping):
        return {str(key): _storage_value(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_storage_value(item) for item in value]
    return value


def _canonical_rows(rows: Iterable[Mapping[str, Any]]) -> list[dict[str, Any]]:
    normalized = [
        {str(key): _storage_value(value) for key, value in row.items()}
        for row in rows
    ]
    return sorted(normalized, key=canonical_content_bytes)


def build_projection_content_manifest(
    table_rows: Mapping[str, Iterable[Mapping[str, Any]]],
) -> tuple[dict[str, dict[str, Any]], str]:
    """Address every projected row and bind the complete table membership."""

    if set(table_rows) != set(PROJECTED_CONTENT_TABLES):
        missing = sorted(set(PROJECTED_CONTENT_TABLES) - set(table_rows))
        extra = sorted(set(table_rows) - set(PROJECTED_CONTENT_TABLES))
        raise ValueError(
            f"Ops Projection content table drift: missing={missing}, extra={extra}"
        )
    manifest: dict[str, dict[str, Any]] = {}
    for table in PROJECTED_CONTENT_TABLES:
        rows = _canonical_rows(table_rows[table])
        manifest[table] = {
            "row_count": len(rows),
            "content_digest": projection_content_digest({"rows": rows}),
        }
    return manifest, projection_content_digest({"tables": manifest})


__all__ = [
    "PROJECTED_CONTENT_TABLES",
    "build_projection_content_manifest",
    "canonical_content_bytes",
    "projection_content_digest",
]
